crates/net/discv4/src/lib.rs

```diff
@@ -578,41 +647,128 @@ impl Discv4Service {
+                record,
+                last_seen: Instant::now(),
+                last_enr_seq: Some(last_enr_seq),
+                fork_id,
+            };
+            let _ = self.kbuckets.insert_or_update(
+                &key,
+                entry,
+                NodeStatus {
+                    state: ConnectionState::Connected,
+                    direction: ConnectionDirection::Outgoing,
+                },
+            );
+        } else {
+            // update the value and re-request the ENR
+            let _ = self.insert_or_update(record, Some(last_enr_seq));
+        }
+    }
+
+    /// Updates the node entry
+    ///
+    /// Inserts the given node into the table.
+    ///
+    /// If EIP-868 is enabled this will send a followup request to ask for the ENR of the node, if
+    /// the given enr seq is `Some`.
+    fn insert_or_update(
+        &mut self,
+        record: NodeRecord,
+        mut last_enr_seq: Option<u64>,
+    ) -> InsertResult<NodeKey> {
+        if record.id == self.local_node_record.id {
+            return InsertResult::Failed(FailureReason::InvalidSelfUpdate)
+        }
+
+        // If EIP868 extension is disabled then we want to ignore this
+        if !self.config.enable_eip868 {
+            last_enr_seq = None;
+        }
+
+        // if the peer included a enr seq in the pong then we can try to request the ENR of that
+        // node
+        let has_enr_seq = last_enr_seq.is_some();
+
         let key = kad_key(record.id);
-        let entry = NodeEntry { record, last_seen: Instant::now() };
-        match self.kbuckets.insert_or_update(
+        let entry = NodeEntry { record, last_seen: Instant::now(), last_enr_seq, fork_id: None };
+        let res = self.kbuckets.insert_or_update(
             &key,
             entry,
             NodeStatus {
                 state: ConnectionState::Connected,
                 direction: ConnectionDirection::Outgoing,
             },
-        ) {
+        );
+
+        match &res {
             InsertResult::Inserted => {
                 debug!(target : "discv4",?record, "inserted new record to table");
                 self.notify(DiscoveryUpdate::Added(record));
+
+                if has_enr_seq {
+                    // request the ENR of the node
+                    self.send_enr_request(record);
+                }
             }
-            InsertResult::ValueUpdated { .. } | InsertResult::Updated { .. } => {
+            InsertResult::ValueUpdated => {
                 trace!(target : "discv4",?record,  "updated record");
+                if has_enr_seq {
+                    // request the ENR of the node
+                    self.send_enr_request(record);
+                }
             }
             InsertResult::Failed(FailureReason::BucketFull) => {
-                debug!(target : "discv4", ?record,  "discovered new record but bucket is full");
-                self.notify(DiscoveryUpdate::Discovered(record));
```

Comment:
Given that we now handle bucket cleaning properly, there is no need to send another Discovered event. This should address @Will-Smith11's concern around unbounded memory in the peer manager, but we haven't tested that yet so would be good to confirm.